Use _StridedShard to replace shard_order field in DTensorSpec

zpcore opened the pull request:
Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167300
* #167887



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk

zpcore committed:
Use _StridedShard to replace shard_order field in DTensorSpec

[ghstack-poisoned]

zpcore commented on torch/csrc/distributed/Placement.h:
```diff
@@ -59,21 +69,21 @@ class StridedShard : public Shard {
 
-  bool operator==(const StridedShard& rhs) const {
-    return dim == rhs.dim && split_factor == rhs.split_factor;
+  // Override virtual equals to handle polymorphic comparison correctly
```
Once _StridedShard is not a subclass of Shard, we can clean this up.

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

Skylion007 commented on torch/distributed/tensor/_redistribute.py:
```diff
@@ -458,6 +456,7 @@ def find_min_cost_path(
                     new_path = path + [next_state]
                     counter += 1
                     heapq.heappush(pq, (new_cost, counter, next_state, new_path))
+        torch.distributed.breakpoint()
```
Leftover debugging?

zpcore commented on test/distributed/tensor/test_utils.py:
```diff
@@ -788,6 +788,7 @@ def world_size(self) -> int:
 
     @with_comms
     def test_StridedShard_to_shard_order(self):
+        torch.manual_seed(0)
```
Good point, I didn't realize this. I think it maybe fine for this test. If developer care about the seed, they will call `torch.manual_seed` in certain tests, so this one won't impact that much.

wconstab replied:
i don't understand why the seed(0) was needed. can you explain?

zpcore commented on torch/distributed/tensor/_redistribute.py:
```diff
@@ -491,9 +490,19 @@ def generate_graph_based_transform_infos(
+                dst_spec.placements, dst_spec.mesh
+            )
+        )
+        assert src_shard_order is not None and dst_shard_order is not None
```
Right, updated

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/Placement.h:
```diff
@@ -44,10 +45,19 @@ class Shard : public Placement {
 
-  bool operator==(const Shard& rhs) const {
+  // Virtual method for polymorphic comparison
+  virtual bool equals(const Shard& rhs) const {
```
do you know if there is much performance impact to using a virtual method here?

an alternative way might be to add one more field to every Shard to identify its class type (e.g. Shard ==0, StridedShard ==1, etc), then avoid having a virtual method.

@swolchok can you comment intuitively if this is worth it or the virtual method is fine?

swolchok replied:
> can you remind me what is included in the 'fast path'

outline of C++ dispatch fast path as of #167475:
- check for custom op handler
- build sharding propagation cache key and check the cache (call out to python sharding prop on cache miss or needs_pytree)
- dispatch to local op if participating, fill in defaults if not. (may call out to python slow path if needed)
- specialized return_and_correct_aliasing implementation for view ops or non-participation, return result. for inplace or out variant ops (and exactly aten::equal.default), call out to python slow path to do their particular quirks.

Many of the slow path things are fixable on demand; just trying to keep the problem well-scoped and incremental.

> does it include any binary ops like aten.add? i'd expect placement to show up in those cases, for example.

For what? uncached sharding propagation is and remains in Python

zpcore commented:
The failing of test `test/distributed/_composable/fsdp/test_fully_shard_clip_grad_norm_.py::TestClipGradNormWorldSize4::test_clip_grad_norm_2d` is real.  Below is my findings:

Original code intentionally make `_StridedShard(dim, sf=...)==Shard(dim)` to bypass issue in _StridedShard sharding propagation.  When I make `_StridedShard(dim, sf=...)!=Shard(dim)` in this PR, it will trigger the redistribution:
```
source placements: (_StridedShard(dim=0, sf=2), Shard(dim=0)) --> target placements: (Shard(dim=0), Shard(dim=0))
redistribution trace: S(0)[1]S(0)[0]->RS(0)->RS(1)->S(0)S(1)->S(0)[0]S(0)[1])
```
While previously, this is a no-op, doesn't need a redistribution.  
Need to think about how to fix this for FSDP.

zpcore commented:
> The failing of test `test/distributed/_composable/fsdp/test_fully_shard_clip_grad_norm_.py::TestClipGradNormWorldSize4::test_clip_grad_norm_2d` is real. Below is my findings:
> 
> Original code intentionally make `_StridedShard(dim, sf=...)==Shard(dim)` to bypass issue in _StridedShard sharding propagation. When I make `_StridedShard(dim, sf=...)!=Shard(dim)` in this PR, it will trigger the redistribution:
> 
> ```
> source placements: (_StridedShard(dim=0, sf=2), Shard(dim=0)) --> target placements: (Shard(dim=0), Shard(dim=0))
> redistribution trace: S(0)[1]S(0)[0]->RS(0)->RS(1)->S(0)S(1)->S(0)[0]S(0)[1])
> ```
> 
> While previously, this is a no-op, doesn't need a redistribution. Need to think about how to fix this for FSDP.

Further investigation:

The `source placements: (_StridedShard(dim=0, sf=2), Shard(dim=0)) --> target placements: (Shard(dim=0), Shard(dim=0))` happens in loss backward when we update the grad for model parameters (called in op like aten.\_foreach_mul_.Tensor()). For some reason, the target model parameters becomes `(Shard(dim=0), Shard(dim=0))` instead of `(_StridedShard(dim=0, sf=2), Shard(dim=0))`. I believe the `_StridedShard` get implicitly cast to `Shard` somehow in the target placement, which should not happen.

zpcore commented:
> > The failing of test `test/distributed/_composable/fsdp/test_fully_shard_clip_grad_norm_.py::TestClipGradNormWorldSize4::test_clip_grad_norm_2d` is real. Below is my findings:
> > Original code intentionally make `_StridedShard(dim, sf=...)==Shard(dim)` to bypass issue in _StridedShard sharding propagation. When I make `_StridedShard(dim, sf=...)!=Shard(dim)` in this PR, it will trigger the redistribution:
> > ```
> > source placements: (_StridedShard(dim=0, sf=2), Shard(dim=0)) --> target placements: (Shard(dim=0), Shard(dim=0))
> > redistribution trace: S(0)[1]S(0)[0]->RS(0)->RS(1)->S(0)S(1)->S(0)[0]S(0)[1])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > While previously, this is a no-op, doesn't need a redistribution. Need to think about how to fix this for FSDP.
> 
> Further investigation:
> 
> The `source placements: (_StridedShard(dim=0, sf=2), Shard(dim=0)) --> target placements: (Shard(dim=0), Shard(dim=0))` happens in loss backward when we update the grad for model parameters (called in op like aten._foreach_mul_.Tensor()). For some reason, the target model parameters becomes `(Shard(dim=0), Shard(dim=0))` instead of `(_StridedShard(dim=0, sf=2), Shard(dim=0))`. I believe the `_StridedShard` get implicitly cast to `Shard` somehow in the target placement, which should not happen.

This is the place where the `_StridedShard` silently becomes `Shard`: https://github.com/pytorch/pytorch/blob/f6a79b2a4ac92a9077fc03cb124e3bc0a0a28730/torch/distributed/tensor/_ops/_pointwise_ops.py#L531

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented:
> @zpcore so i know earlier you came up with an execution plan, but it seems there are still a bunch of delicate number of order of operations issues. Do we need fix the is_shard() situation first to avoid bugs like you saw here?

I think the bug is due to `_StridedShard` is a subclass of `Shard`. `is_shard()` may also cause issue that I haven't seen yet. My plan is to:
- First, after merging this PR, upstream the shard_order user API in https://github.com/pytorch/pytorch/pull/165092
- Second, update `is_shard()`
- (laster some time) make `_StridedShard` a separate class.

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -308,10 +312,12 @@ def _maybe_convert_StridedShard_to_shard_order(
         return tuple(shard_order)
 
-    def _verify_shard_order(self, shard_order: ShardOrder) -> None:
+    def _verify_shard_order(
```
unclear who should call this and why, and why its possible for shard_order to be incorrect in the first place, 

also, like the other staticmethods, seems more like a util in placements module than a staticmethod on spec.

wconstab reviewed (commented):
Somewhere in this PR i think we need to add a long-form 'note' in the code that explains the design of shard_order and StridedShard.

It should cover the way we store the metadata, as well as the way users give it to us and the way we give it back.  This is important to clearly document since it is potentially confusing- we go back and forth between representations and we have functions that return optional types (e.g. `shard_order() -> Optional[ShardOrder]`).  

You can put the note in one file and then you can refer to it from multiple places in the docstring for a function (look around torch codebase for how people use `see Note` - e.g. `aten/src/ATen/CPUGeneratorImpl.cpp: * See Note [Acquire lock when using ran
dom generators]`

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -538,6 +564,12 @@ def device_mesh(self) -> DeviceMesh:
         return self.mesh
 
+    @property
+    def shard_order(self) -> Optional[ShardOrder]:
```
i'm wondering why this returns an optional shard order.  If we can't convert StridedShard to shard, then should we define shard_order as left to right and return that shard order, to be interpreted alongside our placement tuple which includes stridedshard?

zpcore commented on torch/csrc/distributed/Placement.h:
```diff
@@ -59,21 +69,23 @@ class StridedShard : public Shard {
+      return false;
+    }
+    const auto& rhs_strided = static_cast<const StridedShard&>(rhs);
+    return dim == rhs_strided.dim && split_factor == rhs_strided.split_factor;
```
Good catch to reuse! I didn't think about any overhead here...

zpcore commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -73,29 +73,34 @@ class DTensorSpec:
         self._hash: int | None = None
 
+    @staticmethod
+    def normalize_placements_into_shard_order(
```
Let me create a new file `placement_utils.py` to handle those functions.

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented:
> Somewhere in this PR i think we need to add a long-form 'note' in the code that explains the design of shard_order and StridedShard.
> 
> It should cover the way we store the metadata, as well as the way users give it to us and the way we give it back. This is important to clearly document since it is potentially confusing- we go back and forth between representations and we have functions that return optional types (e.g. `shard_order() -> Optional[ShardOrder]`).
> 
> You can put the note in one file and then you can refer to it from multiple places in the docstring for a function (look around torch codebase for how people use `see Note` - e.g. `aten/src/ATen/CPUGeneratorImpl.cpp: * See Note [Acquire lock when using ran dom generators]`

I wrote an outline summary about `Note [Sharding representation in DTensorSpec]` and AI fills in those details pretty well :)

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

As title. 

Also fixes `_StridedShard(dim, sf=...)==Shard(dim)` returns True issue.

Also updates pointwise op to propagate _StridedShard

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -924,9 +924,7 @@ def redistribute(
-        new_spec.shard_order = shard_order
-    else:
-        new_spec.shard_order = ()
+        assert new_spec._maybe_update_placements_given_shard_order(shard_order)
```
what happens if placements contain StridedShard and shard_order is also specified? should we assert this can't happen?

wconstab commented on torch/distributed/tensor/placement_utils.py:
```diff
@@ -0,0 +1,401 @@
+        tensor_dim = entry.tensor_dim
+        mesh_dims = entry.mesh_dims
+        for idx in range(len(mesh_dims)):
+            # TODO(zpcore): split_factor from `view` and `shard order`
```
is this something being addressed in @weifengpy's PR?  I didn't understand what the TODO means

wconstab commented on torch/distributed/tensor/placement_utils.py:
```diff
@@ -0,0 +1,401 @@
+    order by examining the split_factor of each _StridedShard and determining its position
+    in the execution order. If the _StridedShard configuration cannot be represented as a
+    valid ShardOrder (i.e., there's no shard order that produces the observed split_factors),
+    this function returns None.
```
maybe its worth saying when this can happen and when it cant? (example of the view unflatten?)

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Use _StridedShard to replace shard_order field in DTensorSpec"

Remove the `shard_order` field from DTensorSpec. `shard_order` should be represented using `_StridedShard`.

Also moved some DTensorSpec related util functions into new file `placement_utils.py`.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
